_internal/src/types.ts

```diff
@@ -394,13 +394,10 @@ export interface ScopedMutator {
  * @typeParam Data - The type of the data related to the key
  * @typeParam MutationData - The type of the data returned by the mutator
  */
-export type KeyedMutator<Data> = <MutationData>(
-  data?:
-    | MutationData
-    | Promise<MutationData | undefined>
-    | MutatorCallback<MutationData>,
+export type KeyedMutator<Data> = <MutationData = Data>(
+  data?: Data | Promise<Data | undefined> | MutatorCallback<Data>,
```

Comment:
@promer94 Why was this change reverted? It makes the generic completely useless and returning different `MutationData` to be passed to `populateCache` impossible with proper types.

This probably needs to be overload based to function properly, otherwise `MutationData` will just be inferred improperly from the data arg, although it is mostly possible to do without overloads with some rather extensive types:

```ts
type KeyedMutator<Data> = <MutationData = Data>(data?: MutationData | Promise<MutationData | undefined> | MutatorCallback<MutationData>, opts?: boolean | Omit<MutatorOptions<Data, MutationData>, 'populateCache'> & (MutationData extends Data ? Pick<MutatorOptions<Data, Data>, 'populateCache'> : { populateCache: Exclude<MutatorOptions<Data, MutationData>['populateCache'], boolean> })) => Promise<Data | MutationData | undefined>;
```

I have tested the above and it works as expected, except in the case where `opts` is not passed or is a `boolean`. This is why I suggested the overload approach, since if options is not passed as an object, typescript doesn't enforce it being passed.